### ISSUE TITLE
Key Vault: Data plane commands fail behind a proxy

### ISSUE THREAD

**Issue by mdekoster** [Azure/azure-cli#2505]
The data plan commands fail with a timeout error if executed from behind a proxy. It seems the data plane  commands do not use the "HTTP(S)_PROXY" (and/or REQUESTS_CA_BUNDLE) environment setting.

requests.packages.urllib3.connectionpool : Starting new HTTPS connection (1): ****.vault.azure.net
msrest.exceptions : Error occurred in request., ConnectionError: HTTPSConnectionPool(host='****.vault.azure.net', port=443): Max retries exceeded with url: /secrets?api-version=2016-10-01 (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x00000000046173C8>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond',)) Max retries exceeded attempting to connect to vault. The vault may not exist or you may need to flush your DNS cache and try again later.

**Comment by tjprescott** [comment]
Does this only happen with KeyVault or with other data planes as well?

**Comment by mdekoster** [comment]
I have seen it only on the keyvault data plane commands for setting keys and secrets.

**Comment by tjprescott** [comment]
Hi @mdekoster, an update to the KeyVault data plane SDK should resolve this issue. It should be fixed in the next release. As a workaround, you can try installing 'azure-keyvault==0.3.2' directly (assuming you did a pip install).

**Comment by tjprescott** [comment]
Reopening this issue since the change for the dataplane had to be reverted.

**Comment by mdekoster** [comment]
The keyvault dataplane commands are still not proxy aware in the keyvault package version 2.0.8. Can someone solve this issue?

**Comment by tjprescott** [comment]
This is not an issue with the KeyVault CLI commands--it is an issue with the KeyVault SDK.

**Comment by schaabs** [comment]
@mdekoster I tried to reproduce your issue setting http_proxy and https_proxy, however it seemed to route through the proxy correctly.  I confirmed by monitoring the network on both the client and proxy server.  

If you're scripting the CLI to create the vault and then immediately create secrets, keys, etc. I think you might actually be running into issue https://github.com/Azure/azure-cli/issues/3348.  You could confirm if this was in fact the case by adding a wait (10-15 seconds) to your script after creating the vault, or trying the commands in your script manually.

**Comment by mdekoster** [comment]
@schaabs Within the organisation I am currently working for, all traffic is routed through a proxy. The ‘az keyvault secret <command>’, ‘az keyvault key <command>’ and az keyvault certificate <command> result in a https timeout. The commands are done on exsisting keyvaults.

Below the output from the keyvault commands in debug mode. The 'keyvault list' command against the control plane is successful. The 'keyvault secret list' command against the data plane ends in the http timeout.

>>
PS> az keyvault list -o table --debug
Command arguments ['keyvault', 'list', '-o', 'table']
Current cloud config:
{'endpoints': {'active_directory': 'https://login.microsoftonline.com',
               'active_directory_data_lake_resource_id': 'https://datalake.azure.net/',
               'active_directory_graph_resource_id': 'https://graph.windows.net/',
               'active_directory_resource_id': 'https://management.core.windows.net/',
               'batch_resource_id': 'https://batch.core.windows.net/',
               'gallery': 'https://gallery.azure.com/',
               'management': 'https://management.core.windows.net/',
               'resource_manager': 'https://management.azure.com/',
               'sql_management': 'https://management.core.windows.net:8443/',
               'vm_image_alias_doc': 'https://raw.githubusercontent.com/Azure/azure-rest-api-specs/master/arm-compute/qu
ickstart-templates/aliases.json'},
 'is_active': True,
 'name': 'AzureCloud',
 'profile': 'latest',
 'suffixes': {'azure_datalake_analytics_catalog_and_job_endpoint': 'azuredatalakeanalytics.net',
              'azure_datalake_store_file_system_endpoint': 'azuredatalakestore.net',
              'keyvault_dns': '.vault.azure.net',
              'sql_server_hostname': '.database.windows.net',
              'storage_endpoint': 'core.windows.net'}}
Registered application event handler 'CommandTableParams.Loaded' at <function add_id_parameters at 0x03D94780>
Registered application event handler 'CommandTable.Loaded' at <function add_id_parameters at 0x03D94780>
Successfully loaded command table from module 'keyvault'.
Extensions directory: 'C:\Users\...\.azure\cliextensions'
Application event 'CommandTable.Loaded' with event data {'command_table': {'keyvault list': <azure.cli.core.commands.CliCommand object at 0x03D92750>}}
Application event 'CommandParser.Loaded' with event data {'parser': AzCliCommandParser(prog='az', usage=None, description=None, formatter_class=<class 'argparse.HelpFormatter'>, conflict_handler='error', add_help=True)}
Application event 'CommandTableParams.Loaded' with event data {'command_table': {'keyvault list': <azure.cli.core.comman
ds.CliCommand object at 0x03D92750>}}
Application event 'CommandParser.Parsing' with event data {'argv': ['keyvault', 'list', '-o', 'table']}
Application event 'CommandParser.Parsed' with event data {'command': 'keyvault list', 'args': Namespace(_command_package='keyvault', _jmespath_query=None, _log_verbosity_debug=False, _log_verbosity_verbose=False, _output_format='table', _parser=AzCliCommandParser(prog='az keyvault list', usage=None, description='', formatter_class=<class 'argparse.HelpFormat
ter'>, conflict_handler='error', add_help=True), _validators=[], command='keyvault list', func=<azure.cli.core.commands.
CliCommand object at 0x03D92750>, resource_group_name=None, subcommand= [...]
Getting management service client client_type=KeyVaultManagementClient
Using CA bundle file at 'C:\Users\...\.certificates\CertStore.pem'.
attempting to read file C:\Users\...\.azure\accessTokens.json as utf-8-sig
adal-python : 05181479-4f68-4b9f-89dc-8d37eb12ec7e - Authority:Performing instance discovery: https://login.microsoftonline.com/6e93a626-8aca-4dc1-9191-ce291b4b75a1
adal-python : 05181479-4f68-4b9f-89dc-8d37eb12ec7e - Authority:Performing static instance discovery
adal-python : 05181479-4f68-4b9f-89dc-8d37eb12ec7e - Authority:Authority validated via static instance discovery
adal-python : 05181479-4f68-4b9f-89dc-8d37eb12ec7e - TokenRequest:Getting token from cache with refresh if necessary.
adal-python : 05181479-4f68-4b9f-89dc-8d37eb12ec7e - OAuth2Client:finding with query: {"_clientId": "04b07795.....7b46", "userId": "...k.com"}
adal-python : 05181479-4f68-4b9f-89dc-8d37eb12ec7e - OAuth2Client:Looking for potential cache entries:
adal-python : 05181479-4f68-4b9f-89dc-8d37eb12ec7e - OAuth2Client:{"_clientId": "04b07795....7b46", "userId": "...k.com"}
adal-python : 05181479-4f68-4b9f-89dc-8d37eb12ec7e - OAuth2Client:Found 3 potential entries.
adal-python : 05181479-4f68-4b9f-89dc-8d37eb12ec7e - OAuth2Client:Resource specific token found.
adal-python : 05181479-4f68-4b9f-89dc-8d37eb12ec7e - OAuth2Client:Returning token from cache lookup, AccessTokenId: b'Ez
xN6OXYrbbgW6A+4+YE/4ryrVM12SSV/Kb59hAk/gU=', RefreshTokenId: b'urztjD3WrqLP1I8dQMKNWwNj1Ub/35gaW7oCJkWUPyI='
msrest.pipeline : Configuring request: timeout=100, verify='C:\\Users\\...\\.certificates\\CertStore.pem', cert=None
msrest.pipeline : Configuring redirects: allow=True, max=30
msrest.pipeline : Configuring proxies: ''
msrest.pipeline : Evaluate proxies against ENV settings: True
msrest.pipeline : Configuring retry: max_retries=3, backoff_factor=0.8, max_backoff=90
urllib3.connectionpool : Starting new HTTPS connection (1): management.azure.com
urllib3.connectionpool : https://management.azure.com:443 "GET /subscriptions/dadea0f6...255/resources?$filter=resourceType%20eq%20%27Microsoft.KeyVault%2Fvaults%27&api-version=2015-11-01 HTTP/1.1" 200 371
msrest.http_logger : Request URL: 'https://management.azure.com/subscriptions/dadea0f6...255/resources?$filter=resourceType%20eq%20%27Microsoft.KeyVault%2Fvaults%27&api-version=2015-11-01'
msrest.http_logger : Request method: 'GET'
msrest.http_logger : Request headers:
msrest.http_logger :     'User-Agent': 'python/3.6.1 (Windows-10-10.0.14393-SP0) requests/2.18.4 msrest/0.4.14 msrest_az
ure/0.4.13 keyvaultmanagementclient/0.40.0 Azure-SDK-For-Python AZURECLI/2.0.16'
msrest.http_logger :     'Accept-Encoding': 'gzip, deflate'
msrest.http_logger :     'Accept': 'application/json'
msrest.http_logger :     'Connection': 'keep-alive'
msrest.http_logger :     'Authorization': '*****'
msrest.http_logger :     'x-ms-client-request-id': 'd3203eac-9c53-11e7-9328-00155d679a11'
msrest.http_logger :     'CommandName': 'keyvault list'
msrest.http_logger :     'Content-Type': 'application/json; charset=utf-8'
msrest.http_logger :     'accept-language': 'en-US'
msrest.http_logger : Request body:
msrest.http_logger : None
msrest.http_logger : Response status: 200
msrest.http_logger : Response headers:
msrest.http_logger :     'Cache-Control': 'no-cache'
msrest.http_logger :     'Pragma': 'no-cache'
msrest.http_logger :     'Content-Type': 'application/json; charset=utf-8'
msrest.http_logger :     'Content-Encoding': 'gzip'
msrest.http_logger :     'Expires': '-1'
msrest.http_logger :     'Vary': 'Accept-Encoding'
msrest.http_logger :     'x-ms-ratelimit-remaining-subscription-reads': '14994'
msrest.http_logger :     'x-ms-request-id': '75834d8e-9ba1-4e00-a223-20072e3a25e9'
msrest.http_logger :     'x-ms-correlation-request-id': '75834d8e-9ba1-4e00-a223-20072e3a25e9'
msrest.http_logger :     'x-ms-routing-request-id': 'UKSOUTH:20170918T092911Z:75834d8e-9ba1-4e00-a223-20072e3a25e9'
msrest.http_logger :     'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'
msrest.http_logger :     'Date': 'Mon, 18 Sep 2017 09:29:11 GMT'
msrest.http_logger :     'Content-Length': '371'
msrest.http_logger : Response content:
msrest.http_logger : b'{"value":[{"id":"/subscriptions/dadea0f6...rosoft.KeyVault/vaults', 'location': 'westeu [...]
Location    Name         ResourceGroup
----------  -----------  ---------------
westeurope  mdkkv     ..-TEST
westeurope  ..              ..o-KV-CLI
westeurope  ..              ..o-KV-CLI
westeurope  ..v-vm1    ..ovm
westeurope  ..v-vm2    ..ovm




PS > az keyvault secret list --vault-name mdkkv --debug
Command arguments ['keyvault', 'secret', 'list', '--vault-name', 'mdkkv']
Current cloud config:
{'endpoints': {'active_directory': 'https://login.microsoftonline.com',
               'active_directory_data_lake_resource_id': 'https://datalake.azure.net/',
               'active_directory_graph_resource_id': 'https://graph.windows.net/',
               'active_directory_resource_id': 'https://management.core.windows.net/',
               'batch_resource_id': 'https://batch.core.windows.net/',
               'gallery': 'https://gallery.azure.com/',
               'management': 'https://management.core.windows.net/',
               'resource_manager': 'https://management.azure.com/',
               'sql_management': 'https://management.core.windows.net:8443/',
               'vm_image_alias_doc': 'https://raw.githubusercontent.com/Azure/azure-rest-api-specs/master/arm-compute/quickstart-templates/aliases.json'},
'is_active': True,
'name': 'AzureCloud',
'profile': 'latest',
'suffixes': {'azure_datalake_analytics_catalog_and_job_endpoint': 'azuredatalakeanalytics.net',
              'azure_datalake_store_file_system_endpoint': 'azuredatalakestore.net',
              'keyvault_dns': '.vault.azure.net',
              'sql_server_hostname': '.database.windows.net',
              'storage_endpoint': 'core.windows.net'}}
Registered application event handler 'CommandTableParams.Loaded' at <function add_id_parameters at 0x03EFA300>
Registered application event handler 'CommandTable.Loaded' at <function add_id_parameters at 0x03EFA300>
Successfully loaded command table from module 'keyvault'.
Application event 'CommandTable.Loaded' with event data {'command_table': {'keyvault secret list': <azure.cli.core.commands.CliCommand object at 0x03DED930>}}
Application event 'CommandParser.Loaded' with event data {'parser': AzCliCommandParser(prog='az', usage=None, description=None, formatter_class=<class 'argparse.HelpFormatter'>, conflict_handler='error', add_help=True)}
Application event 'CommandTableParams.Loaded' with event data {'command_table': {'keyvault secret list': <azure.cli.core.commands.CliCommand object at 0x03DED930>}}
Application event 'CommandParser.Parsing' with event data {'argv': ['keyvault', 'secret', 'list', '--vault-name', 'mdkkv']}
Application event 'CommandParser.Parsed' with event data {'command': 'keyvault secret list', 'args': Namespace(_command_package='keyvault', _jmespath_query=None, _log_verbosity_debug=False, _log_verbosity_verbose=False, _output_format='json',
_parser=AzCliCommandParser(prog='az keyvault secret list', usage=None, description='List secrets in a specified key vault.  The LIST operation is applicable to the entire vault, however only the base secret identifier and attributes are provided in the response. Individual secret versions are not listed in  [...]
msrest.pipeline : Configuring request: timeout=100, verify=True, cert=None
msrest.pipeline : Configuring redirects: allow=True, max=30
msrest.pipeline : Configuring proxies: ''
msrest.pipeline : Evaluate proxies against ENV settings: True
msrest.pipeline : Configuring retry: max_retries=3, backoff_factor=0.8, max_backoff=90
urllib3.connectionpool : Starting new HTTPS connection (1): mdkkv.vault.azure.net
msrest.exceptions : Error occurred in request., ConnectionError: HTTPSConnectionPool(host='mdkkv.vault.azure.net', port=443): Max retries exceeded with url: /secrets?api-version=2016-10-01 (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x0466BD70>: Failed to establish a new connection: [WinError 10060] A connection attempt
failed because the connected party did not properly respond after a period of time, or established connection failed beca
use connected host has failed to respond',))
Max retries exceeded attempting to connect to vault. The vault may not exist or you may need to flush your DNS cache and
try again later.
<<

**Comment by oddperfect** [comment]
I am running into the same issue when using the key, certificate, and secret commands thru a corporate proxy when on a VPN. I have environment variables HTTP_PROXY and HTTPS_PROXY set and all other azure cli commands work. If I disconnect the VPN and unset the proxy environment variables the key, certificate, and secret commands then work. 

When going thru the VPN, I can curl to the key list URL (but obviously get an unauthorized response) but that shows it is not a firewall blocking the URL.

**Comment by RandalliLama** [comment]
@tjprescott Can you assign this to @schaabs please?

**Comment by oddperfect** [comment]
I wrote a quick python script to do a GET to the key list URL for my keyvault and it was able to make a connection using the proxy. I did this to verify that python 'requests' library was grabbing the correct proxy settings for my setup and it was.

**Comment by schaabs** [comment]
The issue here appears to be that the CLI was using an older version of the Key Vault SDK (0.3.4) which didn't honor session settings when authenticating.  This was fixed a few months back by this change https://github.com/Azure/azure-sdk-for-python/pull/1244 in the SDK but we didn't update the CLI to use the newer version.  Unfortunately, when I first tried to repro this I had a poluted environment which had the later version of azure-keyvault.  The referenced pull request should fix this issue.
